pkg/receive/multitsdb.go

```diff
@@ -0,0 +1,297 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package receive
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage/tsdb"
+	terrors "github.com/prometheus/prometheus/tsdb/errors"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/component"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+	"github.com/thanos-io/thanos/pkg/shipper"
+	"github.com/thanos-io/thanos/pkg/store"
+	"golang.org/x/sync/errgroup"
+)
+
+type MultiTSDB struct {
+	dataDir         string
+	logger          log.Logger
+	reg             prometheus.Registerer
+	tsdbCfg         *tsdb.Options
+	tenantLabelName string
+	labels          labels.Labels
+	bucket          objstore.Bucket
+
+	mtx     *sync.RWMutex
+	tenants map[string]*tenant
+}
+
+type tenant struct {
+	tsdbCfg *tsdb.Options
+
+	readyS *tsdb.ReadyStorage
+	fs     *FlushableStorage
+	s      *store.TSDBStore
+	ship   *shipper.Shipper
+
+	mtx *sync.RWMutex
+}
+
+func newTenant(tsdbCfg *tsdb.Options) *tenant {
+	return &tenant{
+		tsdbCfg: tsdbCfg,
+		readyS:  &tsdb.ReadyStorage{},
+		mtx:     &sync.RWMutex{},
+	}
+}
+
+func (t *tenant) readyStorage() *tsdb.ReadyStorage {
+	return t.readyS
+}
+
+func (t *tenant) store() *store.TSDBStore {
+	t.mtx.RLock()
+	defer t.mtx.RUnlock()
+	return t.s
+}
+
+func (t *tenant) shipper() *shipper.Shipper {
+	t.mtx.RLock()
+	defer t.mtx.RUnlock()
+	return t.ship
+}
+
+func (t *tenant) flushableStorage() *FlushableStorage {
+	t.mtx.RLock()
+	defer t.mtx.RUnlock()
+	return t.fs
+}
+
+func (t *tenant) set(tstore *store.TSDBStore, fs *FlushableStorage, ship *shipper.Shipper) {
+	t.readyS.Set(fs.Get(), int64(2*time.Duration(t.tsdbCfg.MinBlockDuration).Seconds()*1000))
+	t.mtx.Lock()
+	t.fs = fs
+	t.s = tstore
+	t.ship = ship
+	t.mtx.Unlock()
+}
+
+func NewMultiTSDB(
```

Comment:
Nit: would be nice to document all of these exported types with comments, but we can do in a follow up